js/perf/viewer/src/main.js

```diff
@@ -0,0 +1,192 @@
+    }));
+    return elapsedOrNulls[0] !== null ? median(elapsedOrNulls) : null;
+  };
+
+  const getChartData = (testName: string) =>
+    Promise.all(perfData.map(pd => getElapsed(testName, pd)));
+
+  const testChartData = await Promise.all(testNames.map(getChartData));
+  for (let i = 0; i < testNames.length; i++) {
+    chartDatasets.set(testNames[i], testChartData[i]);
+  }
+
+  render();
+}
+
+// Returns the history of perf data with their git revisions, from oldest to newest.
+async function getPerfHistory(ds: Dataset): Promise<[Struct[], string[]]> {
+  const perfData = [], gitRevs = [];
+
+  for (let head = await ds.head(), i = 0; head && i < MAX_PERF_HISTORY; i++) {
+    const val = head.value;
+    invariant(val instanceof Struct);
+    perfData.push(val);
+    gitRevs.push(val.nomsRevision);
+
+    const parentRef = await head.parents.first(); // TODO: how to deal with multiple parents?
+    head = parentRef ? await parentRef.targetValue(ds.database) : null;
+  }
+
+  return [perfData, gitRevs];
+}
+
+// Returns a map of URL param key to value.
+function getParams(): {[key: string]: string} {
+  // Note: this way anything after the # will end up in `params`, which is what we want.
+  const params = {};
+  const paramsIdx = location.href.indexOf('?');
+  if (paramsIdx > -1) {
+    decodeURIComponent(location.href.slice(paramsIdx + 1)).split('&').forEach(pair => {
+      const [k, v] = pair.split('=');
+      params[k] = v;
+    });
+  }
+  return params;
+}
+
+async function render() {
+  if (!chartDatasets) {
+    return;
+  }
+
+  const datasets = [];
+  for (const [testName, elapsed] of chartDatasets) {
+    const [backgroundColor, borderColor] = await genLightAndDarkColors(testName);
+    datasets.push({
+      backgroundColor,
+      borderColor,
+      borderWidth: 1,
+      data: elapsed,
+      label: testName,
+    });
+  }
+
+  // $FlowIssue: Chart is in modules, not imported, so Flow doesn't know about it.
+  new Chart(document.getElementById('chart'), {
+    type: 'line',
+    data: {
+      labels: chartLabels,
+      datasets,
+    },
+    options: {
+      scales: {
+        yAxes: [{
+          scaleLabel: {
+            display: true,
+            labelString: 'elapsed (seconds)',
+          },
+          ticks: {
+            beginAtZero: true,
+          },
+        }],
+        xAxes: [{
+          scaleLabel: {
+            display: true,
+            labelString: 'github.com/attic-labs/noms git revision',
+          },
+        }],
+      },
+    },
+  });
+}
+
+// Returns the median of numbers in `nums`.
+function median(nums: number[]): number {
+  const sorted = nums.slice();
+  sorted.sort();
+  const lenDiv2 = Math.floor(nums.length / 2);
+  let res = nums[lenDiv2];
+  if (nums.length % 2 === 0) {
+    res += nums[lenDiv2 - 1];
+    res /= 2;
+  }
+  return res;
+}
+
+// Generates a light and dark version of some color randomly (but stable) derived from `str`.
```

Comment:
Why do we want this to be random in the first place. It seems better to hard code these so we can use our color palette.